File: modules/expo-mlkit-translate/ios/MlkitTranslateModule.swift
import ExpoModulesCore
// There is no single umbrella `MLKit` Swift module for these APIs — each pod
// ships its own framework module:
//   MLKitTranslate  → Translator, TranslatorOptions, TranslateLanguage,
//                     TranslateRemoteModel, ModelManager.downloadedTranslateModels
//   MLKitLanguageID → LanguageIdentification
//   MLKitCommon     → ModelManager, ModelDownloadConditions,
//                     ModelDownloadUserInfoKey, the download notifications
import MLKitCommon
import MLKitLanguageID
import MLKitTranslate

struct TranslateArgs: Record {
  @Field var text: String = ""
  @Field var source: String = ""
  @Field var target: String = ""
  @Field var allowDownload: Bool = false
}

struct DownloadArgs: Record {
  @Field var language: String = ""
  @Field var requireWifi: Bool = true
}

final class UnsupportedLanguageException: GenericException<String> {
  override var reason: String { "ML Kit cannot translate \"\(param)\"." }
  override var code: String { "ERR_MLKIT_UNSUPPORTED_LANGUAGE" }
}

final class ModelMissingException: GenericException<String> {
  override var reason: String { "The on-device model for \"\(param)\" is not downloaded yet." }
  override var code: String { "ERR_MLKIT_MODEL_MISSING" }
}

final class TranslateFailedException: GenericException<String> {
  override var reason: String { param }
  override var code: String { "ERR_MLKIT_FAILED" }
}

/// Ceiling on a model download.
///
/// ML Kit reports downloads through NotificationCenter, and a download that
/// never posts either terminal notification would leave the promise unsettled
/// forever — which froze the language picker, because its rows stay disabled
/// while a download is in flight. Every download is therefore bounded.
private let downloadTimeout: TimeInterval = 180

public class MlkitTranslateModule: Module {

  /// A cached translator plus the bookkeeping that keeps it alive while in use.
  ///
  /// Dropping a translator mid-translate makes that call fail, so eviction
  /// cannot simply discard it: a busy entry is unlinked from the cache and
  /// marked, and the last in-flight call lets it go.
  private final class Entry {
    let translator: Translator
    var inUse: Int = 0
    var dropWhenIdle: Bool = false
    init(_ translator: Translator) { self.translator = translator }
  }

  /// Loaded translators keyed "source>target".
  ///
  /// Google documents a live translator as occupying 30–150MB of RAM, so this
  /// is deliberately tiny: a chat translates one direction at a time, so every
  /// message after the first reuses an instance.
  private var translators: [String: Entry] = [:]
  private var recency: [String] = []
  private let maxTranslators = 2
  private let lock = NSLock()

  private lazy var languageIdentifier = LanguageIdentification.languageIdentification()
  private lazy var modelManager = ModelManager.modelManager()

  /// BCP-47 tag → ML Kit language, or nil when unsupported.
  private func toMlkit(_ tag: String) -> TranslateLanguage? {
    let wanted = tag.lowercased()
    return TranslateLanguage.allLanguages().first { $0.rawValue.lowercased() == wanted }
  }

  private func acquire(source: TranslateLanguage, target: TranslateLanguage) -> Entry {
    lock.lock()
    defer { lock.unlock() }

    let key = "\(source.rawValue)>\(target.rawValue)"
    let entry: Entry
    if let existing = translators[key] {
      entry = existing
      recency.removeAll { $0 == key }
    } else {
      let options = TranslatorOptions(sourceLanguage: source, targetLanguage: target)
      entry = Entry(Translator.translator(options: options))
      translators[key] = entry
    }
    recency.append(key)
    entry.inUse += 1

    while recency.count > maxTranslators {
      let oldest = recency.removeFirst()
      if oldest == key { recency.append(oldest); break }  // never evict what we just took
      guard let evicted = translators.removeValue(forKey: oldest) else { continue }
      if evicted.inUse > 0 { evicted.dropWhenIdle = true }
    }
    return entry
  }

  private func release(_ entry: Entry) {
    lock.lock()
    entry.inUse -= 1
    lock.unlock()
  }

  private func closeAll() {
    lock.lock()
    translators.removeAll()
    recency.removeAll()
    lock.unlock()
  }

  private func isDownloaded(_ language: TranslateLanguage) -> Bool {
    let model = TranslateRemoteModel.translateRemoteModel(language: language)
    return modelManager.isModelDownloaded(model)
  }

  public func definition() -> ModuleDefinition {
    Name("MlkitTranslateModule")

    // Hand back the 30–150MB the translators hold when the app leaves the front.
    OnAppEntersBackground { self.closeAll() }
    OnDestroy { self.closeAll() }

    Function("getSupportedLanguages") { () -> [String] in
      TranslateLanguage.allLanguages().map { $0.rawValue }
    }

    AsyncFunction("identifyLanguage") { (text: String, promise: Promise) in
      self.languageIdentifier.identifyLanguage(for: text) { code, error in
        if let error = error {
          promise.reject(TranslateFailedException(error.localizedDescription))
          return
        }
        promise.resolve(code ?? "und")
      }
    }

    AsyncFunction("isModelDownloaded") { (language: String, promise: Promise) in
      guard let lang = self.toMlkit(language) else { promise.resolve(false); return }
      promise.resolve(self.isDownloaded(lang))
    }

    AsyncFunction("downloadModel") { (args: DownloadArgs, promise: Promise) in
      guard let lang = self.toMlkit(args.language) else {
        promise.reject(UnsupportedLanguageException(args.language)); return
      }
      let model = TranslateRemoteModel.translateRemoteModel(language: lang)
      if self.modelManager.isModelDownloaded(model) { promise.resolve(nil); return }

      let conditions = ModelDownloadConditions(
        allowsCellularAccess: !args.requireWifi,
        allowsBackgroundDownloading: true
      )

      // The iOS SDK reports completion through NotificationCenter rather than a
      // callback, so listen for this model's terminal event and settle exactly
      // once — including on a timeout, so a stalled download can never leave the
      // picker spinning forever.
      let settleLock = NSLock()
      var didSettle = false
      var observers: [NSObjectProtocol] = []
      var timeoutWork: DispatchWorkItem?

      func finish(_ block: @escaping () -> Void) {
        settleLock.lock()
        if didSettle { settleLock.unlock(); return }
        didSettle = true
        settleLock.unlock()

        timeoutWork?.cancel()
        observers.forEach { NotificationCenter.default.removeObserver($0) }
        observers.removeAll()
        block()
      }

      let matches: (Notification) -> Bool = { note in
        guard let downloaded = note.userInfo?[ModelDownloadUserInfoKey.remoteModel.rawValue]
                as? TranslateRemoteModel else { return false }
        return downloaded.language == lang
      }

      observers.append(NotificationCenter.default.addObserver(
        forName: .mlkitModelDownloadDidSucceed, object: nil, queue: nil
      ) { note in
        guard matches(note) else { return }
        finish { promise.resolve(nil) }
      })
      observers.append(NotificationCenter.default.addObserver(
        forName: .mlkitModelDownloadDidFail, object: nil, queue: nil
      ) { note in
        guard matches(note) else { return }
        let error = note.userInfo?[ModelDownloadUserInfoKey.error.rawValue] as? Error
        finish {
          promise.reject(TranslateFailedException(
            error?.localizedDescription ?? "Model download failed."))
        }
      })

      let work = DispatchWorkItem {
        // The download may still land later; the model check on the next attempt
        // will pick it up. What matters is that the caller is unblocked now.
        finish {
          promise.reject(TranslateFailedException(
            "Model download timed out after \(Int(downloadTimeout))s."))
        }
      }
      timeoutWork = work
      DispatchQueue.main.asyncAfter(deadline: .now() + downloadTimeout, execute: work)

      self.modelManager.download(model, conditions: conditions)
    }

    AsyncFunction("deleteModel") { (language: String, promise: Promise) in
      guard let lang = self.toMlkit(language) else {
        promise.reject(UnsupportedLanguageException(language)); return
      }
      let model = TranslateRemoteModel.translateRemoteModel(language: lang)
      self.modelManager.deleteDownloadedModel(model) { error in
        if let error = error {
          promise.reject(TranslateFailedException(error.localizedDescription))
          return
        }
        promise.resolve(nil)
      }
    }

    AsyncFunction("getDownloadedModels") { (promise: Promise) in
      promise.resolve(self.modelManager.downloadedTranslateModels.map { $0.language.rawValue })
    }

    AsyncFunction("translate") { (args: TranslateArgs, promise: Promise) in
      guard let source = self.toMlkit(args.source) else {
        promise.reject(UnsupportedLanguageException(args.source)); return
      }
      guard let target = self.toMlkit(args.target) else {
        promise.reject(UnsupportedLanguageException(args.target)); return
      }
      if source == target { promise.resolve(args.text); return }

      let entry = self.acquire(source: source, target: target)
      let settleLock = NSLock()
      var didSettle = false
      func settle(_ block: () -> Void) {
        settleLock.lock()
        if didSettle { settleLock.unlock(); return }
        didSettle = true
        settleLock.unlock()
        self.release(entry)
        block()
      }

      func run() {
        entry.translator.translate(args.text) { result, error in
          if let error = error {
            settle { promise.reject(TranslateFailedException(error.localizedDescription)) }
            return
          }
          settle { promise.resolve(result ?? args.text) }
        }
      }

      if args.allowDownload {
        // Explicit opt-in (the picker screen). Wi-Fi-only so a 30MB pull never
        // lands on someone's data plan without them asking for it.
        let conditions = ModelDownloadConditions(
          allowsCellularAccess: false, allowsBackgroundDownloading: true)
        entry.translator.downloadModelIfNeeded(with: conditions) { error in
          if let error = error {
            settle { promise.reject(TranslateFailedException(error.localizedDescription)) }
            return
          }
          run()
        }
        return
      }

      // Chat path: never block a bubble on a 30MB download. ML Kit pivots
      // through English, so BOTH endpoints must already be present.
      for lang in [source, target] where lang != .english {
        if !self.isDownloaded(lang) {
          settle { promise.reject(ModelMissingException(lang.rawValue)) }
          return
        }
      }
      run()
    }
  }
}

File: modules/expo-mlkit-translate/android/src/main/java/expo/modules/mlkittranslate/MlkitTranslateModule.kt
package expo.modules.mlkittranslate

import com.google.mlkit.common.model.DownloadConditions
import com.google.mlkit.common.model.RemoteModelManager
import com.google.mlkit.nl.languageid.LanguageIdentification
import com.google.mlkit.nl.languageid.LanguageIdentifier
import com.google.mlkit.nl.translate.TranslateLanguage
import com.google.mlkit.nl.translate.TranslateRemoteModel
import com.google.mlkit.nl.translate.Translation
import com.google.mlkit.nl.translate.Translator
import com.google.mlkit.nl.translate.TranslatorOptions
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicReference

class TranslateArgs : Record {
  @Field var text: String = ""
  @Field var source: String = ""
  @Field var target: String = ""
  @Field var allowDownload: Boolean = false
}

class DownloadArgs : Record {
  @Field var language: String = ""
  @Field var requireWifi: Boolean = true
}

internal class UnsupportedLanguageException(tag: String) :
  CodedException("ERR_MLKIT_UNSUPPORTED_LANGUAGE", "ML Kit cannot translate \"$tag\".", null)

internal class ModelMissingException(tag: String) :
  CodedException(
    "ERR_MLKIT_MODEL_MISSING",
    "The on-device model for \"$tag\" is not downloaded yet.",
    null,
  )

internal class TranslateFailedException(cause: Throwable?) :
  CodedException("ERR_MLKIT_FAILED", cause?.message ?: "ML Kit translation failed.", cause)

class MlkitTranslateModule : Module() {

  /**
   * A cached translator plus the bookkeeping that keeps it alive while in use.
   *
   * Closing a translator mid-translate makes that call fail, so eviction cannot
   * simply call close(): a busy entry is unlinked from the cache and marked, and
   * the last in-flight call closes it on the way out.
   */
  private class Entry(val translator: Translator) {
    var inUse: Int = 0
    var closeWhenIdle: Boolean = false
  }

  /**
   * Loaded translators, keyed "source>target".
   *
   * Google documents a live translator as occupying 30–150MB of RAM, so these
   * are NOT kept per language pair. A chat translates one direction at a time,
   * so a tiny cache gets every message after the first for free.
   */
  private val translators = LinkedHashMap<String, Entry>()
  private val maxTranslators = 2

  private val languageIdentifier: LanguageIdentifier by lazy {
    LanguageIdentification.getClient()
  }

  private val modelManager: RemoteModelManager by lazy { RemoteModelManager.getInstance() }

  /** BCP-47 tag → ML Kit's internal constant, or null when unsupported. */
  private fun toMlkit(tag: String): String? = TranslateLanguage.fromLanguageTag(tag)

  @Synchronized
  private fun acquire(source: String, target: String): Entry {
    val key = "$source>$target"
    val entry = translators.remove(key) ?: Entry(
      Translation.getClient(
        TranslatorOptions.Builder()
          .setSourceLanguage(source)
          .setTargetLanguage(target)
          .build(),
      ),
    )
    translators[key] = entry                 // re-insert = most recently used
    entry.inUse += 1

    while (translators.size > maxTranslators) {
      val oldestKey = translators.keys.first()
      if (oldestKey == key) break            // never evict what we just acquired
      val evicted = translators.remove(oldestKey) ?: continue
      if (evicted.inUse == 0) evicted.translator.close() else evicted.closeWhenIdle = true
    }
    return entry
  }

  @Synchronized
  private fun release(entry: Entry) {
    entry.inUse -= 1
    if (entry.inUse <= 0 && entry.closeWhenIdle) entry.translator.close()
  }

  @Synchronized
  private fun closeAll() {
    translators.values.forEach {
      if (it.inUse == 0) it.translator.close() else it.closeWhenIdle = true
    }
    translators.clear()
  }

  override fun definition() = ModuleDefinition {
    Name("MlkitTranslateModule")

    // Free the translators (and their 30–150MB) when the app goes away.
    OnActivityDestroys { closeAll() }
    OnDestroy { closeAll() }

    Function("getSupportedLanguages") {
      TranslateLanguage.getAllLanguages()
    }

    AsyncFunction("identifyLanguage") { text: String, promise: Promise ->
      languageIdentifier.identifyLanguage(text)
        .addOnSuccessListener { code -> promise.resolve(code ?: "und") }
        .addOnFailureListener { error -> promise.reject(TranslateFailedException(error)) }
    }

    AsyncFunction("isModelDownloaded") { language: String, promise: Promise ->
      val tag = toMlkit(language)
      if (tag == null) { promise.resolve(false); return@AsyncFunction }
      modelManager.isModelDownloaded(TranslateRemoteModel.Builder(tag).build())
        .addOnSuccessListener { downloaded -> promise.resolve(downloaded) }
        .addOnFailureListener { promise.resolve(false) }
    }

    AsyncFunction("downloadModel") { args: DownloadArgs, promise: Promise ->
      val tag = toMlkit(args.language)
      if (tag == null) {
        promise.reject(UnsupportedLanguageException(args.language)); return@AsyncFunction
      }
      val conditions = DownloadConditions.Builder().apply {
        if (args.requireWifi) requireWifi()
      }.build()
      // A Task can only settle once, but guard anyway: a promise resolved twice
      // is a hard crash in Expo, and this one is driven by callbacks.
      val settled = AtomicBoolean(false)
      modelManager.download(TranslateRemoteModel.Builder(tag).build(), conditions)
        .addOnSuccessListener { if (settled.compareAndSet(false, true)) promise.resolve(null) }
        .addOnFailureListener { error ->
          if (settled.compareAndSet(false, true)) promise.reject(TranslateFailedException(error))
        }
    }

    AsyncFunction("deleteModel") { language: String, promise: Promise ->
      val tag = toMlkit(language)
      if (tag == null) {
        promise.reject(UnsupportedLanguageException(language)); return@AsyncFunction
      }
      modelManager.deleteDownloadedModel(TranslateRemoteModel.Builder(tag).build())
        .addOnSuccessListener { promise.resolve(null) }
        .addOnFailureListener { error -> promise.reject(TranslateFailedException(error)) }
    }

    AsyncFunction("getDownloadedModels") { promise: Promise ->
      modelManager.getDownloadedModels(TranslateRemoteModel::class.java)
        .addOnSuccessListener { models -> promise.resolve(models.map { it.language }) }
        .addOnFailureListener { error -> promise.reject(TranslateFailedException(error)) }
    }

    AsyncFunction("translate") { args: TranslateArgs, promise: Promise ->
      val source = toMlkit(args.source)
      val target = toMlkit(args.target)
      if (source == null) {
        promise.reject(UnsupportedLanguageException(args.source)); return@AsyncFunction
      }
      if (target == null) {
        promise.reject(UnsupportedLanguageException(args.target)); return@AsyncFunction
      }
      if (source == target) { promise.resolve(args.text); return@AsyncFunction }

      val entry = acquire(source, target)
      val settled = AtomicBoolean(false)
      fun settle(block: () -> Unit) {
        if (settled.compareAndSet(false, true)) { release(entry); block() }
      }

      fun run() {
        entry.translator.translate(args.text)
          .addOnSuccessListener { result -> settle { promise.resolve(result) } }
          .addOnFailureListener { error -> settle { promise.reject(TranslateFailedException(error)) } }
      }

      if (args.allowDownload) {
        // Explicit opt-in (the picker screen). Wi-Fi-only so a 30MB pull never
        // lands on someone's data plan without them asking for it.
        entry.translator.downloadModelIfNeeded(DownloadConditions.Builder().requireWifi().build())
          .addOnSuccessListener { run() }
          .addOnFailureListener { error -> settle { promise.reject(TranslateFailedException(error)) } }
        return@AsyncFunction
      }

      // Chat path: never block a bubble on a 30MB download. Report the miss and
      // let JS decide (it kicks off the picker's download instead).
      //
      // ML Kit pivots through English, so BOTH endpoints must be present. The
      // checks run concurrently, so the counter and the first-miss slot are
      // atomic rather than plain vars — ML Kit happens to call these back on the
      // main looper today, but nothing in the API promises that.
      val needed = listOf(source, target).filter { it != TranslateLanguage.ENGLISH }
      if (needed.isEmpty()) { run(); return@AsyncFunction }
      val remaining = AtomicInteger(needed.size)
      val missing = AtomicReference<String?>(null)
      val failure = AtomicReference<Throwable?>(null)
      needed.forEach { tag ->
        modelManager.isModelDownloaded(TranslateRemoteModel.Builder(tag).build())
          .addOnSuccessListener { downloaded ->
            if (!downloaded) missing.compareAndSet(null, tag)
            if (remaining.decrementAndGet() == 0) {
              val absent = missing.get()
              when {
                absent != null -> settle { promise.reject(ModelMissingException(absent)) }
                failure.get() != null -> settle { promise.reject(TranslateFailedException(failure.get())) }
                else -> run()
              }
            }
          }
          .addOnFailureListener { error ->
            failure.compareAndSet(null, error)
            if (remaining.decrementAndGet() == 0) {
              val absent = missing.get()
              if (absent != null) settle { promise.reject(ModelMissingException(absent)) }
              else settle { promise.reject(TranslateFailedException(error)) }
            }
          }
      }
    }
  }
}
